Qstnrs

bogdan-cornianu opened the pull request:
Implement the Questionnaires app.

bogdan-cornianu committed:
create skeleton for questionnaires app.

bogdan-cornianu committed:
remove qstnrs.

bogdan-cornianu committed:
change name from qstnrs to questionnaires.

bogdan-cornianu committed:
remove questionnaire project as the app named tests conflicts with the tests python package.

bogdan-cornianu committed:
add index, questions and pages templates; improve model.

bogdan-cornianu committed:
add search functionality for models in the admin interface.

bogdan-cornianu committed:
redirect to index.html on /

bogdan-cornianu committed:
replace check of empty list with the '{% empty %}' tag

bogdan-cornianu committed:
refactor to use render() instead of using separate template loader and context objects.

bogdan-cornianu committed:
add ability to go to the next or previous page and display results at the end of the questionnaire.

bogdan-cornianu committed:
add logic to display result for questionnaire.

bogdan-cornianu committed:
add files for packaging; draft for new result calculation.

bogdan-cornianu committed:
added gap folder to .gitignore

bogdan-cornianu committed:
improve logic to select possible improvements.

bogdan-cornianu committed:
improve logic to compute improvements for the results page.

bogdan-cornianu committed:
improve result template to display the suggested improvements.

bogdan-cornianu committed:
add information on how to dump database to a json file in order to use it as a fixture in tests; add shell script to run unit tests and show code coverage.

bogdan-cornianu committed:
add database dump to .gitignore

bogdan-cornianu committed:
delete temporary file MANIFEST.in~

bogdan-cornianu committed:
improve logic to calculate improvements.

bogdan-cornianu committed:
add unit tests.

bogdan-cornianu committed:
improve unit tests.

bogdan-cornianu committed:
sort descending the unselected answers.

bogdan-cornianu committed:
fix unit tests.

bogdan-cornianu committed:
add Intellij Idea project files to git ignore.

bogdan-cornianu committed:
add requirements.txt

bogdan-cornianu committed:
add information on how to install the Questionnaire egg file; add autodiscover for packages to setup.py.

bogdan-cornianu committed:
small code cleanup.

bogdan-cornianu committed:
add html files for 404 and 500 errors; add support for pagination

bogdan-cornianu committed:
fix declaration for required packages.

bogdan-cornianu committed:
add tox configuration file.

bogdan-cornianu committed:
fix pagination for questionnaires list.

bogdan-cornianu committed:
remove old migrations.

bogdan-cornianu committed:
add pytest; modify tox to work woth pytest; fix dependencies in setup.py

bogdan-cornianu committed:
create migration for removing the result table.

bogdan-cornianu committed:
add simple test.

bogdan-cornianu committed:
add initial migration.

bogdan-cornianu committed:
add South.

bogdan-cornianu committed:
fix algorithm to generate user improvements

bogdan-cornianu committed:
improve templates.

bogdan-cornianu committed:
add new packages.

bogdan-cornianu committed:
tests for urls and utility functions.

bogdan-cornianu committed:
remove test coverage script.

bogdan-cornianu committed:
Add tests for views.

bogdan-cornianu committed:
Remove redundant index url

The view to call for the index page is already set in the app's
urls.py

bogdan-cornianu committed:
Remove requirement for coverage module

bogdan-cornianu committed:
Comment out tests in tests.py

All tests are available under questionnaire/tests and should
be run with pytest

bogdan-cornianu committed:
Add default text for result description

bogdan-cornianu committed:
Improve get_score_for(user_choices) to use QuerySet

bogdan-cornianu committed:
Add fixture to use with pytest

bogdan-cornianu committed:
Add migrations to remove the Result table

Migration 0002 - adds result description and result upper_limit
columns to the Questionnaire table.

Migration 0003 - create Result table.

Migration 0004 (data migration) - copy data from Result table into the
corresponding fields in the Questionnaire table.

Migration 0005 - delete the Result table.

Migration 0006 - delete the result description and result upper_limit
columns from the Questionnaire table

bogdan-cornianu committed:
Remove result description and upper limit fields
from the Questionnaire model.

bogdan-cornianu committed:
Improve unit tests

* Remove id from conftest.py tables definition.
* Replace hard-coded lists with query sets.
* Explicit import instead of *.
* Replace hard-coded urls with Django's reverse() function.

bogdan-cornianu committed:
Refactor utils.py

* Replace answers_for_questionnaire(questionnaire_id) with a QuerySet.

* Refactor get_suggestions_for(questionnaire_id, user_choices, better)
  to use QuerySets.

* Remove test for answers_for_questionnaire(questionnaire_id).

bogdan-cornianu committed:
* Replace hard-coded urls with the url tag.
* Add maximum score information to the result page.

bogdan-cornianu committed:
* Add name to url patterns.
* Add protection for KeyError and ValueError in views.
* Replace import * with specific imports.
* Replace hard-coded urls with reverse().
* Change check of non-existing previous page id to -1 instead of 0
  in the page view.
* Add max_score to the result view.
* Add get_max_score_for(questionnaire_id) to calculate the maximum
  score the user could obtain.

bogdan-cornianu committed:
Create README.md

bogdan-cornianu committed:
Update README.md

bogdan-cornianu committed:
Additional ordering for answers with same score

* Order by answer_text answers which have the same score.

bogdan-cornianu committed:
Add test for maximum score of a questionnaire.

bogdan-cornianu committed:
Add tests for string in url.

bogdan-cornianu committed:
Merge branch 'qstnrs' of https://github.com/bogdan-cornianu/dlp into qstnrs

laurafeier commented on qstnrs/README.rst:
```diff
@@ -0,0 +1,23 @@
+Quick start
+-----------
+1. Install the Questionnaire egg file by running:
+    easy_install Questionnaire.egg
```
Why do you prefer easy_install over pip?

bogdan-cornianu replied:
I do not prefer one over the other. 
I found out about easy_install when I was reading about creating and installing egg files (http://peak.telecommunity.com/DevCenter/PythonEggs).
I'll change it to use pip.

laurafeier commented on qstnrs/pytest.ini:
```diff
@@ -0,0 +1,4 @@
+[pytest]
+DJANGO_SETTINGS_MODULE = qstnrs.settings
+addopts = --reuse-db
```
Why is --reuse-db required?

bogdan-cornianu replied:
--reuse-db is required because I don't think it's necessary to create the test database every time the tests run.

laurafeier replied:
How do you ensure the migrations are correct if you are not executing them?

bogdan-cornianu replied:
You're right about that. I'll remove --reuse-db.

laurafeier commented on qstnrs/qstnrs/urls.py:
```diff
@@ -0,0 +1,10 @@
+urlpatterns = patterns(
+    '',
+    url(r'^qstnrs/', include('questionnaire.urls')),
+    url(r'^admin/', include(admin.site.urls)),
```
I think it's better to allow projects that use this app to make the decision where they want the django admin urls to live. Since the decision weather the django admin app is installed lies in the django project I think it's better to make the urls live there too.

laurafeier replied:
You can check out Projects vs. apps -> https://docs.djangoproject.com/en/1.8/intro/tutorial01/#creating-models

bogdan-cornianu replied:
Qstnrs is the django project and Questionnaire is the app, so the django admin urls already live in the django project. Isn't this correct?

laurafeier replied:
I'm sorry, I was confused by the multiple admin.autodiscover(). I think it should be only in one place. You have it in the app urls as well. You should drop it since it might not work if django admin app is not installed in your project.

bogdan-cornianu replied:
Yes, I'll delete it from the app urls.

laurafeier commented on qstnrs/questionnaire/forms.py:
```diff
@@ -0,0 +1,16 @@
+        super(PageForm, self).__init__(*args, **kwargs)
+
+        for question in page.question_set.all():
+            answers = [(answer.id, answer.answer_text)
```
Why not use values_list? This way you will not load all objects data in memory.

laurafeier commented on qstnrs/questionnaire/forms.py:
```diff
@@ -0,0 +1,16 @@
+        page = kwargs.pop('page')
+        super(PageForm, self).__init__(*args, **kwargs)
+
+        for question in page.question_set.all():
```
I think you can use values_list here as well: page.question_set.values_list('id', 'question_text') and then you can fetch the answers Answer.objects.values_list('id', 'answer_text'). Be aware of the fact that this will generate a query for each question. If however your db has many question, you might get performance issues. In this case you might consider iterating through something like this: page.question_set.values('id','question_text', 'answer', 'answer__answer_text')

bogdan-cornianu replied:
ok.

laurafeier commented on qstnrs/questionnaire/migrations/0004_result_desc_result_upper_limit.py:
```diff
@@ -0,0 +1,68 @@
+            result_list = questionnaire.result_set.all()
+            if len(result_list) > 0:
+                questionnaire.result_description = result_list[0].description
+                questionnaire.result_upper_limit = result_list[0].upper_limit
```
I don't think you need to store the upper limit. Upper limit is considered to be the max score one can have by selecting all the right answers, right?

laurafeier replied:
Nevermind, I saw the last migration.

laurafeier commented on qstnrs/questionnaire/migrations/0003_auto__add_result__chg_field_questionnaire_result_upper_limit__chg_fiel.py:
```diff
@@ -0,0 +1,79 @@
+
+    def forwards(self, orm):
+        # Adding model 'Result'
+        db.create_table('questionnaire_result', (
```
I don't think this is working. This is the second table with this name 'questionnaire_result'. You already created one in the first migration and never deleted it. I think you can drop this one.

bogdan-cornianu replied:
Yes, you're right.
I'll remove the table creation from this migration.

laurafeier commented on qstnrs/questionnaire/models.py:
```diff
@@ -0,0 +1,42 @@
+    def __unicode__(self):
+        return self.question_text[:100]
+
+    class Meta:
```
Why not use https://docs.djangoproject.com/en/1.8/ref/models/options/#order-with-respect-to?

bogdan-cornianu replied:
I will use that and write a schema migration to remove the page_order and question_order fields.

laurafeier commented on qstnrs/questionnaire/views.py:
```diff
@@ -0,0 +1,130 @@
+    return render(request, "page.html", {
+        "form": form,
+        "page_has_questions": page_has_questions,
+        "previous_page_id": previous_page_id,
```
I think it will be much more readable if you provide something like 'has_next': previous_page_id > -1, 'has_prev': next_page_id > -1. It will make the template code much more easier to read.

bogdan-cornianu replied:
I agree, I'll make the necessary changes.

laurafeier commented on qstnrs/questionnaire/utils.py:
```diff
@@ -0,0 +1,65 @@
+
+def get_score_for(user_choices):
+    """Calculate the user's total score based on selected answers."""
+    return sum(list(Answer.objects.filter(id__in=user_choices).values_list(
```
There's another way: Answer.objects.aggregate(total=Sum('answer_score'))

bogdan-cornianu replied:
ok.

laurafeier commented on qstnrs/questionnaire/utils.py:
```diff
@@ -0,0 +1,65 @@
+    return possible_answers
+
+
+def on_same_page(answer, possible_choices):
```
You should move this function before the function where it's used. It makes everything much more clearer for the one that's reading the code.

bogdan-cornianu replied:
ok.

laurafeier commented on qstnrs/questionnaire/views.py:
```diff
@@ -0,0 +1,130 @@
+    try:
+        first_page = Page.objects.filter(questionnaire_id=questionnaire_id). \
+            order_by('page_order')[0].id
+        url = reverse('qstnrs-page', args=(questionnaire_id, str(first_page)))
```
You should drop the str. Your url pattern already accepts only numbers.

bogdan-cornianu replied:
ok.

laurafeier commented on qstnrs/questionnaire/views.py:
```diff
@@ -0,0 +1,130 @@
+        if request.method == 'POST':
+            form = PageForm(request.POST, page=current_page)
+            if form.is_valid():
+                if 'choices' not in request.session:
```
you can also use session.setdefault and drop the if statement

bogdan-cornianu replied:
ok.

laurafeier commented on qstnrs/questionnaire/views.py:
```diff
@@ -0,0 +1,130 @@
+        worse_suggestions = get_suggestions_for(
+            int(questionnaire_id), user_choices, better=False
+        )
+    except ValueError:
```
Since your url pattern only accepts numbers for questionnaire_id, is it possible for the int() cast to raise ValueError? Does the pattern allow None value for questionnaire_id? In this case is ValueError raised?

bogdan-cornianu replied:
It's not possible for the int() cast to raise ValueError since the url pattern accepts only numbers.
The pattern doesn't allow None values, it requires at least one number, ValueError is not raised.
I'll remove the try block.

laurafeier commented on qstnrs/tests/conftest.py:
```diff
@@ -0,0 +1,233 @@
+
+    # create pages
+    Page.objects.create(
+        questionnaire_id=1,
```
How can one know to which of the created questionnaire this page belongs to? since the id is auto-generated 

https://docs.djangoproject.com/en/1.8/ref/models/relations/

laurafeier commented on qstnrs/tests/conftest.py:
```diff
@@ -0,0 +1,233 @@
+    """Get unselected choices for the first questionnaire"""
+    return Answer.objects.filter(question__page__questionnaire_id=1).exclude(
+        answer_text__in=['Vlad', '12', 'Blue', 'Lily', 'Dogs', 'Mercedes',
+                         'Pizza', 'Drunken Rat']).values_list('id', flat=True)
```
pep8: No newline at the end of the file?

bogdan-cornianu replied:
Sorry about this, I'll fix it.

bogdan-cornianu committed:
Use pip to install the egg package instead of easy_install.

bogdan-cornianu committed:
Remove --reuse-db from pytest.ini

Remove --reuse-db to ensure migrations are run.

bogdan-cornianu committed:
Remove creating the Result table since it was
created during the initial migration.

bogdan-cornianu committed:
Remove unused import of admin from django.contrib.

bogdan-cornianu committed:
Get all answers for a question by using values_list.

bogdan-cornianu committed:
* Use order_with_respect_to for ordering instead of
a custom order column for the Page and Question
tables.

* Add migration to remove the page_order and
question_order columns from Page and Question
tables.

bogdan-cornianu committed:
Send has_previous and has_next flags to the template
to know when to render the Previous, Next, Show Results
buttons and the Back to Questionnaire list link.

bogdan-cornianu committed:
Use aggregate to calculate the score for a user
and the maximum score for a questionnaire.

bogdan-cornianu committed:
* Remove unnecessary try blocks.
* Remove unnecessary int() and str() casts.
* Move on_same_page() before the place it is used.

bogdan-cornianu committed:
Remove manually setting of id when creating
pages, questions and answers.

bogdan-cornianu committed:
Add new line at the end of file.

laurafeier commented on qstnrs/tests/conftest.py:
```diff
@@ -0,0 +1,85 @@
+        question_text="Where do you go for lunch?")
+
+    # create answers
+    name_qstn.answer_set.create(answer_text="Vlad", answer_score=-1)
```
How about using bulk_create?

bogdan-cornianu replied:
ok.

bogdan-cornianu committed:
Create the answers using bulk_create.

bogdan-cornianu committed:
Improve retrieval of question and answers for a page.

bogdan-cornianu committed:
Use setdefault() for self.fields.
